Better user-key documentation directions (#171)
Onemind-Services-LLC/netbox-secrets: README.md

* `display_setting`
  * __Type__: `Dict`
  * __Description__: Set display setting for concrete model
  * __Default__: `{}`
  * __Options__: `{'app.model': 'display_default'}`
  * __Example__: `{'dcim.device': 'full_width_page', 'virtualization.virtualmachine': 'right_page'}`
* `enable_contacts`
  * __Type__: `Boolean`
  * __Description__: Enable contacts for secret
  * __Default__: `False`
* `public_key_size`
  * __Type__: `Integer`
  * __Description__: Size of the public key
  * __Default__: `2048`
  * __Options__: `2048`, `4096`, `8192`
* `top_level_menu`
  * __Type__: `Boolean`
  * __Description__: Enable top level menu
  * __Default__: `False`

## Extra Configuration

The following options are inherited from NetBox to configure the cookies:

* `SESSION_COOKIE_SECURE`
  * __Type__: `Boolean`
  * __Description__: [Session Cookie Secure](https://docs.netbox.dev/en/stable/configuration/security/#session_cookie_secure)
* `LOGIN_TIMEOUT`
  * __Type__: `Integer`
  * __Description__: [Login Timeout](https://docs.netbox.dev/en/stable/configuration/security/#login_timeout)

__Note: These options should be set in the NetBox configuration file.__

## User Keys

> Each user within NetBox can associate his or her account with an RSA public key. If activated by an administrator, this user key will contain a unique, encrypted copy of the AES master key needed to retrieve secret data.

Reviewer: ```suggestion
Each user within NetBox can associate their account with an RSA public key. If activated by an administrator, this user key will contain a unique, encrypted copy of the AES master key needed to retrieve secret data.
```

I think it's also better to mention activation is one time

> User keys may be created by users individually, however they are of no use until they have been activated by a user who already possesses an active user key.

Reviewer: This sounds a bit redundant to me. It's better to rephrase it

Author: This was copied from `docs/models/userkey.md`. I personally Prefer every usage or deployment related inside the main `Readme.md` file and only development documentation inside the `docs` directory.

I will bring everything into the `Readme.md` file and you tell me what you think of it

## Pynetbox

The `pynetbox` Python lib can be used to interact with the NetBox API as well. Here are some examples of how to use `pynetbox` to interact with the NetBox Secrets plugin.

For more information on `pynetbox`, please refer to the [pynetbox documentation](https://pynetbox.readthedocs.io/en/latest/).

### Generating RSA Key Pair using Pynetbox

You can use the `<netbox_url>/api/plugins/secrets/generate-rsa-key-pair/` endpoint to generate a new RSA key pair. The keys are returned in PEM format.

```json
{
    "public_key": "<public key>",
    "private_key": "<private key>"
}
```

To generate a new RSA key pair, using `pynetbox`, you can use the following code:

```python
import pynetbox

nb = pynetbox.api("netbox_url", token="your_netbox_token")

key_pair = nb.plugins.secrets.generate_rsa_key_pair.create()
```

The `key_pair` variable will contain the public and private keys in PEM format.

### Activating User Key using Pynetbox

To activate a user key, you can use the `<netbox_url>/api/plugins/secrets/activate-user-key/` endpoint. The endpoint requires the private key in PEM format.

```json
{

>     "private_key": "<private key>"

Author: okay

}
```

To activate a user key, using `pynetbox`, you can use the following code:

```python
import pynetbox

nb = pynetbox.api("netbox_url", token="your_netbox_token")

nb.plugins.secrets.activate_user_key.create(private_key="<private_key>")
```

The user key will be activated and can now be used to encrypt and decrypt secrets.

### Generating Session Key using Pynetbox

To generate a session key, you can use the `<netbox_url>/api/plugins/secrets/session-keys/` endpoint. The endpoint requires the private key in PEM format.

```json
{

>     "private_key": "<private key>"

Reviewer 2: I noticed that it doesn't mention the optional `preserve_key` field in the payload. This field allows users to reuse an existing session key instead of creating a new one.

```

The `session_key` variable will contain the session key. The session key can now be used to encrypt and decrypt secrets.

### Creating Secrets using Pynetbox

To create a secret, you can use the `<netbox_url>/api/plugins/secrets/secrets/` endpoint. The endpoint requires the session key in the `X-Session-Key` header. The data sent needs to be in the following format:

```json
{
  "assigned_object_type": "string",
  "assigned_object_id": 2147483647,
  "role": 1, // Role ID
  "name": "string",
  "plaintext": "string",
}
```

To create a secret, using `pynetbox`, you can use the following code:

```python
import pynetbox
import requests

nb = pynetbox.api("netbox_url", token="your_netbox_token")

# retrieving the session key
session_key = nb.plugins.secrets.session_keys.create(private_key="<private_key>")
session_key = dict(session_key)["session_key"]

# updating the session key in the headers
session = requests.Session()
session.headers.update({"X-Session-Key": session_key})
nb.http_session = session

# creating the secret
secret = {
    "assigned_object_type": "dcim.device",
    "assigned_object_id": 2147483647,
    "role": 1,
    "name": "admin",
    "plaintext": "foobar",
}

> nb.plugins.secrets.secrets.create(data=secret, session=session)

Reviewer 3: Sorry for stepping in but I found an error calling the pynetbox API here:

Using this line from the example, I got 
```
ynetbox.core.query.RequestError: The request failed with code 400 Bad Request: {'assigned_object_type': ['This field is required.'], 'assigned_object_id': ['This field is required.'], 'role': ['This field is required.'], 'plaintext': ['This field is required.']}
```
After trying around a bit, I got it working by using this call instead `nb.plugins.secrets.secrets.create(secret)` because create() is defined as `create(*args, **kwargs)`, taking the properties of the objects to be created directly, or a list of them as dicts. See https://pynetbox.readthedocs.io/en/latest/endpoint.html#pynetbox.core.endpoint.Endpoint.create

Nevertheless this documentation enhancement is awesome.

Author: Good one!

I've not tested everything, just put on what I thought would work. I'll take some time to test everything up and update the pull request.

Thanks for the correction!

  "assigned_object_id": 2147483647,
  "role": {
    "name": "string",
    "slug": "8M0SQnM8j0Y3Bto4c7SEs3w_2bvxF1SK9UFFRVSUfzW5CdD12Ro3o-ecA7yJP3OkHgvSy1zlsV84DgKNh2J02-578Q4Tvg0KsOe"
  },
  "name": "string",
  "plaintext": "string",
  "description": "string",
  "comments": "string",
  "tags": [
    {
      "name": "string",
      "slug": "3iRRVSvfqln",
      "color": "392511"
    }
  ],
  "custom_fields": {
    "additionalProp1": "string",
    "additionalProp2": "string",
    "additionalProp3": "string"
  }
}
```

To update a secret, using `pynetbox`, you can use the following code:

```python

import pynetbox
import requests

nb = pynetbox.api("netbox_url", token="your_netbox_token")

# retrieving the session key

session_key = nb.plugins.secrets.session_keys.create(private_key="<private_key>")
session_key = dict(session_key)["session_key"]

# updating the session key in the headers
session = requests.Session()
session.headers.update({"X-Session-Key": session_key})
nb.http_session = session

# updating the secret
secret = {
    "assigned_object_type": "dcim.device",
    "assigned_object_id": 2147483647,
    "role": 1,
    "name": "admin",
    "plaintext": "foobar",
}

> nb.plugins.secrets.secrets.update(id=1, data=secret, session=session)

Reviewer 2: The `update` method requires the `id` to be included in the payload rather than as a parameter. Here is the correct payload example:
```
secret = {
    "id": 1,
    "assigned_object_type": "dcim.device",
    "assigned_object_id": 35,
    "role": 1,
    "name": "admin",
    "plaintext": "foobar",
}
nb.plugins.secrets.secrets.update([secret])
```
Make sure to pass the payload in this format when updating an instance.

If we want to update a single instance, we first need to fetch the instance and then update it like this:
```
x = nb.dcim.devices.get(1)
x.update({
    "name": "test-switch2",
    "serial": "ABC321",
})
```
